### PR TITLE
Remove usage of subpackages from TriBITS Kokkos build and downstream Trilinos packages (#11545)

### PR DESCRIPTION
**Pull request by bartlettroscoe**
CC: @trilinos/compadre, @trilinos/intrepid2, @trilinos/kokkos-kernels, @trilinos/kokkos, @trilinos/muelu, @trilinos/panzer, @trilinos/phalanx, @trilinos/sacado, @trilinos/shylu, @trilinos/stk, @trilinos/stokhos, @trilinos/teuchos, @trilinos/tpetra, @trilinos/trilinoscouplings, @trilinos/xpetra 

## Description

This PR takes a straightforward approach to remove the usage of subpackages from the TriBITS build of Kokkos.

Some of the more significant changes made in this PR include:

* No more Kokkos subpackages `Core`, `Containers`, `Algorithms`, or `Simd` (and all of the associated changes that go with that).
* When enabling Kokkos, all four of the subpackage/subdirectories `core`, `containers`, `algorithms`, and `simd` are always enabled all the time as well as all of their targets and tests.
* Names of many of the Kokkos tests changed from `Kokkos<Subpkg>_XXX` to `Kokkos_<Subpkg>XXX`.
* All downstream `Dependencies.cmake` files now list `Kokkos` and not `KokkosCore`, `KokkosContainers`, etc.
* All optional dependency vars `<Package>_ENABLE_Kokkos<Spkg>` are now just one var `<Package>_ENABLE_Kokkos` (and all of the changes associated with that including user cache vars `<Package>_ENABLE_Kokkos<Spkg>` now becoming `<Package>_ENABLE_Kokkos`).
* All macros in configured `<Package>_Confg.h[.in]` files with the suffixes `_KOKKOSCORE`, `_KOKKOSCONTAINERS`, etc. are now just `_KOKKOS` and this change is propagated to all C++ source files as well.  (This touched a lot of source files.)
* The names of disabled Kokkos tests by the Trilinos GenConfig files, ATDM Trilinos, and other files are updated (performed by a script).

Any of the above could break backward compatibility for downstream software depending on what they depend on.  Below are **some cases where this PR breaks backward compatibility for Trilinos customers**:

1. The user configuring Trilinos has Kokkos enabled but has disabled support for `KokkosContainers` in `Muelu` by passing in `MueLu_ENABLE_KokkosContainers=OFF`.  That can be fixed by changing it to `MueLu_ENABLE_Kokkos=OFF`

2. Downstream (even non-CMake-based) software depends on the `#define` macros coming from Trilinos configured header files.  For example, if downstream C++ software had ifndefs like `#ifdef HAVE_SACADO_KOKKOSCORE`, then that code would be broken.  That would be fixed by changing it to  `#ifdef HAVE_SACADO_KOKKOS`

See the commit log messages for more details.

**NOTES:**

1. The refactorings in the Kokkos package itself was all done manually but the bulk of the other refactorings were done using scripts provided in this email.   (Will provide more detailed notes later.)

3. There is some more cleanup that could be performed in the various packages to remove duplication and redundant code (as will be obvious if you look over the changes carefully) but the code as is should work correctly.

4. This technically breaks backward comparability for both downstream TriBITS packages and downstream non-TriBITS cmake projects.  But it is likely that most downstream non-TriBITS CMake projects likely will not notice any changes since they likely just call `find_package(Kokkos)` or `find_package(Trilinos)` and, for example, to contents of the configured file `KokkosCore_Config.h` are unchanged after this refactoring.

5. It is possible to provide for a little more backward compatibility like maintaining the macro suffixes `_KOKKOSCORE`, `_KOKKOSCONTAINERS`, etc. that would not require changes to source files in Trilinos or source files in downstream projects (through included `<Package>_Config.h` files) but this would require a lot of manual CMake code in the impacted `<packageDir>/CMakeLists.txt` files.

## Tasks

* [x] Wait for the snapshot of Kokkos and KokkosKernels for release-candidate-4.0.01 (see https://github.com/trilinos/Trilinos/pull/11808#issuecomment-1522076279) ... **PR #11817 is merged**.
* [x] Rebase this PR branch on top of 'develop'
* [x] Create PRs for the Kokkos and KokkosKernels changes to their external repos on 'develop' branches:
   - See https://github.com/kokkos/kokkos/pull/6104
   - See https://github.com/kokkos/kokkos-kernels/pull/1817
* [x] Get Trilinos developers to review all of the changes and note the breaks in backward compatibility
* [x] Merge related PRs at same time:
  * [x] Merge this PR #11808 to Trilinos 'develop'
  * ~~[ ] Merge https://gitlab-ex.sandia.gov/trilinos-devops-consolidation/code/GenConfig/-/merge_requests/92 (test name changes)~~ ... **Skipped, see [below](https://github.com/trilinos/Trilinos/pull/11808#issuecomment-1540978902)**
  * ~~[ ] Merge https://gitlab-ex.sandia.gov/trilinos-project/son-ini-files/-/merge_requests/3 (test name changes)~~ ... **Skipped, see [below](https://github.com/trilinos/Trilinos/pull/11808#issuecomment-1540978902)**
* [ ] Get feedback from Trilinos customers that do daily testing against 'develop' to see if they get any breakages

### CODE REVIEW

**Review by lucbv — APPROVED** [review]
The changes made to Kokkos Kernels and MueLu look good to me!

**Comment by bartlettroscoe** [comment]
CC: @ccober6, @crtrott, @srajama1, @jwillenbring 

@trilinos/compadre, @trilinos/intrepid2, @trilinos/kokkos-kernels, @trilinos/kokkos, @trilinos/muelu, @trilinos/panzer, @trilinos/phalanx, @trilinos/sacado, @trilinos/shylu, @trilinos/stk, @trilinos/stokhos, @trilinos/teuchos, @trilinos/tpetra, @trilinos/trilinoscouplings, @trilinos/xpetra,

There are changes in the PR to your packages related to the refactoring to remove subpackages from Kokkos.  This technically breaks the backward compatibility of your packages but it unlikely there are any customers that would experience any breakages (see comments [above](https://github.com/trilinos/Trilinos/pull/11808#issue-1681529946)).

NOTE: @lucbv has already approved this PR so I could just merge it but I wanted to give other developers a chance to take a look before I merge this.  But I am happy to back out this merge if we discover if downstream customers experience problems with this.

**Comment by alanw0** [comment]
@bartlettroscoe thanks for the heads-up. It looks good to me, I'll bring the stk changes to the sierra repository.

File: packages/sacado/cmake/Sacado_config.h.in
Comment (rppawlo) [inline packages/sacado/cmake/Sacado_config.h.in:35]: probably don't need this second define anymore

**Review by rppawlo — APPROVED** [review]
Phalanx and Panzer changes look good to me! Thanks @bartlettroscoe!

**Review by CamelliaDPG — APPROVED** [review]
The Intrepid2 change looks fine, too.  Thanks, @bartlettroscoe!

File: packages/sacado/cmake/Sacado_config.h.in
Comment (bartlettroscoe) [reply packages/sacado/cmake/Sacado_config.h.in:35]: @rppawlo, yes, that is the additional clean I was referring to in the commit message [`37a0259` (#11808)](https://github.com/trilinos/Trilinos/pull/11808/commits/37a025906abb990e0f36d669648e773b23dbd206) and https://github.com/trilinos/Trilinos/pull/11808/files#diff-834265fa841a516cc293104d8f66988af1600be94c7010e3e5adda2a36629afc.

It would be hard to write a script to robustly automatically remove this type of duplication but we can always to a later manual cleanup.

Do you think we should remove all of this duplication manually before merging this PR or can we do it in a later PR?  I was hopping to get this merged and have downstream customers like EMPIRE, Albany, Xyce, Sierra, SPARC, Nalu and others test against this to see if there are major problems in case we need to back this out.  (I added a task for a follow-up PR to to cleanup this duplication in #11545.)

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe we use to have a code owner file that required folks from various team to approve PRs that touch their code base before we could merge. I wonder what happened to that?

@CamelliaDPG, don't know.  You would need to ask someone on the @trilinos/framework team (CC: @sebrowne).  Code owners needing to approve a PR would be nice.

File: packages/sacado/cmake/Sacado_config.h.in
Comment (rppawlo) [reply packages/sacado/cmake/Sacado_config.h.in:35]: separate pr is fine for removing duplication.

**Comment by mayrmt** [comment]
@bartlettroscoe 

>Code owners needing to approve a PR would be nice.

See #11812.

**Comment by ndellingwood** [comment]
@bartlettroscoe please hold off on merge until matching PRs go into kokkos and kokkos-kernels
@lucbv @crtrott we need to coordinate timing of this PR merging, I suggest after snapshot of the patch release release-candidate-4.0.01

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe please hold off on merge until matching PRs go into kokkos and kokkos-kernels @lucbv @crtrott we need to coordinate timing of this PR merging, I suggest after snapshot of the patch release release-candidate-4.0.01

@ndellingwood, holding off is fine.  Who will create the PRs for the Kokkos and KokkosKernels repo?  From https://github.com/trilinos/Trilinos/issues/11752#issuecomment-1492567115 and 42b8e6de0116bdbc5cb66681894416016063b91a I think I have the info I need to pull off the changes for Kokkos and KokkosKernels and apply them to the Kokkos and KokkosKernels repo branches.  So l let me know and I should be able to produce at least the PRs against 'develop' in the Kokkos and KokkosKernels repos.  (And then someone can cherry-pick the commits to the release branch.)

In the meantime, I will work on Kokkos changes on a follow-on branch to make the Kokkos native non-TriBITS CMake build system work with Trilinos (both as an external package and an internal package).

**Comment by bartlettroscoe** [comment]
FYI: I added the label `AT: WIP` and pushed the commit [`21cc39a` (#11808)](https://github.com/trilinos/Trilinos/pull/11808/commits/21cc39a826bda19ffc1f67ba087ab6fe80368c95) so now this PR can't be pushed without another round of testing and approvals.  And depending how long it takes to snapshot in the Kokkos and KokkosKernels release release-candidate-4.0.01, I may end up pushing more commits to get Trilinos working with the native Kokkos CMake build system.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe please hold off on merge until matching PRs go into kokkos and kokkos-kernels @lucbv @crtrott we need to coordinate timing of this PR merging, I suggest after snapshot of the patch release release-candidate-4.0.01

@ndellingwood, thinking about this a little, there are two options for coordinating these changes and Kokkos & KokkosKernels release release-candidate-4.0.01:

* **option-1:**
  - Merge this PR #11808 branch to Trilinos 'develop'
  - Pull off the commits for Kokkos and KokkosKerenls and apply them to the native GitHub repos and the release-candidate-4.0.01 branches in those repos
  - Snapshot the updated release-candidate-4.0.01 branch for Kokkos and KokkosKernels to Trilinos 'develop'

* **option-2:**
  - Snapshot the updated release-candidate-4.0.01 branch for Kokkos and KokkosKernels to Trilinos 'develop'  **without removing Kokkos subpackages** on that branch or in those repos
  - Rebase this PR #11808 branch on top of Trilinos 'develop'
  - Pull off the commits for Kokkos and KokkosKerenls and apply them to the native GitHub repos 'develop' branch (and any up-coming minor releases for Kokkos)

Which option do you do you want to go with?  It sounds like bringing in the release-candidate-4.0.01 branch for Kokkos and KokkosKernels is more urgent so perhaps we should go with **option-2**?  In that case, forget this PR #11808 exists for now and just snapshot in release-candidate-4.0.01 for Kokkos and KokkosKernels.

**Comment by ndellingwood** [comment]
@bartlettroscoe thanks for the response! Regarding options, I'd prefer option 2 as we are ready for snapshotting the patch release and I think option 2 will be the most expedient path.
If you have time can you submit the PRs of the kokkos and kokkos-kernels changes of this PR to their resp. develop branches? If my previous snapshot debacle causes problems let me know and I can help out to save you some time (I would likely just manually copy changes over and issue PRs, my git skills aren't sufficient for handling these messier situations)

**Comment by ndellingwood** [comment]
@bartlettroscoe thanks for all the work in this PR, these updates will be very useful!

**Comment by ndellingwood** [comment]
> Just provide a comment here once you are done snapshotting the release-candidate-4.0.01 branch for Kokkos and KokkosKernels to Trilinos 'develop' without removing Kokkos subpackages. Then I will take care of the rest.

@bartlettroscoe sounds good, thank you!

**Comment by kuberry** [comment]
@bartlettroscoe Can you roll back the changes to packages/compadre/CMakeLists.txt? All of these uses of KokkosCore are for how Compadre uses an existing install of Trilinos (not for how Compadre installs as a package in Trilinos). They will have to change eventually (version dependent), but those changes aren't needed in this PR for Compadre to build with KokkosCore removed as a subpackage. We should keep the changes you made in packages/compadre/cmake/Dependencies.cmake, though.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe Can you roll back the changes to packages/compadre/CMakeLists.txt?

@kuberry, okay, I see that lines 55 through 402 are for the raw CMake build of Compadre.  I will add a commit to remove those changes.

File: packages/kokkos/algorithms/unit_tests/CMakeLists.txt
Comment (dalg24) [inline packages/kokkos/algorithms/unit_tests/CMakeLists.txt:167]: This changes the name of the test (missing underscore)

Comment (dalg24) [reply packages/kokkos/algorithms/unit_tests/CMakeLists.txt:167]: (not a blocker just saying)

Comment (bartlettroscoe) [reply packages/kokkos/algorithms/unit_tests/CMakeLists.txt:167]: Given that subpackages are being removed, some of the tests (like `UnitTest`) had conflicting names because the prefixes became the same `Kokkos_`.  The approach I took was to change from `Kokkos<Subpkg>_<remainder>` to `Kokkos_<Subpkg><remainder>`.  But if you would instead prefer changing this to `Kokkos_<Subpkg>_<remainder>`, we can do that as well.  Just let me know how you would like to names these tests consistently.

**Comment by bartlettroscoe** [comment]
> > Just provide a comment here once you are done snapshotting the release-candidate-4.0.01 branch for Kokkos and KokkosKernels to Trilinos 'develop' without removing Kokkos subpackages. Then I will take care of the rest.
> 
> @bartlettroscoe sounds good, thank you!


@ndellingwood, now that is is complete with the merge of PR:

* #11817

can we pull the trigger on merging this PR and moving the changes to the Kokkos repo?

(I just need to push a small commit to address a native Kokkos CMake build issue before merge this PR.)

File: packages/kokkos/algorithms/unit_tests/CMakeLists.txt
Comment (dalg24) [reply packages/kokkos/algorithms/unit_tests/CMakeLists.txt:167]: Current version is good enough

**Comment by ndellingwood** [comment]
@bartlettroscoe thanks for your work on this!

> can we pull the trigger on merging this PR and moving the changes to the Kokkos repo?

@dalg24 from the review discussion it looks like the naming convention is fine, any other objections to moving forward getting this PR into Trilinos (along with resp. patches to Kokkos and KokkosKernels)?

**Comment by bartlettroscoe** [comment]
With the pushes of the commits 84c23a49645517b8c7cd0844ecc9de80338e8599 and 7c5d7aef7d374f5baf1158f5133f0f91e133626 (which fixes the raw Kokkos CMake build), I believe that this PR is ready to finish review, testing, and merging.

Therefore, I removed the label `AT: WIP` and added the label `AT: RETEST`.

**Comment by bartlettroscoe** [comment]
After rebasing on 'develop' and pushing the new commits 73d0362 and 12a9832 (which fixes the raw Kokkos CMake build), I believe that this PR is ready to finish review, testing, and merging.

Therefore, I removed the label `AT: WIP` and added the label `AT: RETEST`.

File: packages/kokkos/simd/CMakeLists.txt
Comment (dalg24) [inline packages/kokkos/simd/CMakeLists.txt:1]: Technically you need to update the name of the unit test in `packages/kokkos/simd/unit_tests/CMakeLists.txt` too

Comment (bartlettroscoe) [reply packages/kokkos/simd/CMakeLists.txt:1]: @dalg24, I considered that but the existing test in that directory is called `UnitTest_SIMD` and it seemed silly to call it `SimdUnitTest_SIMD`.  To make things more consistent with the other renamed tests, should we call this `SimdUntiTest`?

Comment (dalg24) [reply packages/kokkos/simd/CMakeLists.txt:1]: I can live with the current version that is proposed.  One thing that worries me a bit (I do not expect that can be resolved in this PR) is the number of cmake/ctest scripts and .ini config spec files that explicitly name Kokkos test names.  This is brittle as Kokkos will reserve the right to rename them in the future.

Comment (bartlettroscoe) [reply packages/kokkos/simd/CMakeLists.txt:1]: > This is brittle as Kokkos will reserve the right to rename them in the future.

@dalg24, but that is true of any test in any Trilinos package.  If a test gets renamed but not the disable cache var then the worst thing that will happen is that the test will be re-enabled and will either pass or fail.  If it fails, someone will hopefully notice the name change and then update the *.ini file.

It would be really nice if all of these packages could define more robust test suites and disable expensive tests by default in full debug builds (i.e. `CMAKE_BUILD_TYPE=Debug`).  That would mostly eliminate the need to be disabling tests in automated builds.  (And users would not see so many test failures when they built Trilinos.)

**Comment by bartlettroscoe** [comment]
FYI: The last PR testing iteration [above](https://github.com/trilinos/Trilinos/pull/11808#issuecomment-1534127949) failed due to a problem with the rebase not replacing KokkosCore with Kokkos in the muelu/cmake/Dependencies.cmake files.  That is fixed with commit bdb73b00400a2f89ce067ed82934b01414d2c381.   I also noticed a duplicate Kokkos in different dependency lists from Stokhos that I fixed in commit 05bd114292bcd11388c4f8a90fac1c95954257ac.

The next set of PR builds should pass.

**Review by crtrott — DISMISSED** [review]
Gonna block this until the Kokkos Core /KokkosKernels changes merge on the primary repos. Currently the Kokkos Core PR fails.

**Comment by bartlettroscoe** [comment]
> Gonna block this until the Kokkos Core /KokkosKernels changes merge on the primary repos. Currently the Kokkos Core PR fails.

I will take a look and see why those are failing.  (I my local testing, I was not enabling tests in the native Kokkos CMake build.)

**Comment by bartlettroscoe** [comment]
FYI: With the latest commits bec68bf0341ada03c999fa9c352fb9c6c5fdc0d7 and bcdb13779079c9dc5755b4ff580bcdb410da1dfb, I verified that building Kokkos with the native build system by enabling all of the tests and examples with:

```
-D Kokkos_ENABLE_TESTS=ON \
-D Kokkos_ENABLE_BENCHMARKS=ON \
-D Kokkos_ENABLE_EXAMPLES=ON \
```

gives 100% passing tests:

```
100% tests passed, 0 tests failed out of 32
```

NOTE: The commit bec68bf0341ada03c999fa9c352fb9c6c5fdc0d7 is identical to the commit [kokkos/kokkos@`469d46b` (#6104)](https://github.com/kokkos/kokkos/pull/6104/commits/469d46b61d528902d40692e91c6603289095b6fe) in the native Kokkos repo.  (The latter is on top of Kokkos 'develop' which is much newer than the version of Kokkos currently snapshotted into Trilinos 'develop' that commit bec68bf0341ada03c999fa9c352fb9c6c5fdc0d7 is based off of.  That is why commit bcdb13779079c9dc5755b4ff580bcdb410da1dfb was needed to fix the native Kokkos CMake test build in this PR.)

**Comment by bartlettroscoe** [comment]
> Gonna block this until the Kokkos Core /KokkosKernels changes merge on the primary repos. Currently the Kokkos Core PR fails.

@crtrott , with the issue with the Kokkos changes in https://github.com/kokkos/kokkos/pull/6104 resolved with https://github.com/kokkos/kokkos/pull/6104/commits/4c8215b945478e1192a8f937f4007fd21c72cd74, can we approve this PR #11808 so we can get this merged and see how this impacts close Trilinos customers?

I am hopping that the breaks in backward compatibility will not actually impact real customers (because they are unlikely to have configure options or usage patterns that depended on Kokkos being broken into subpackages).

**Comment by bartlettroscoe** [comment]
@crtrott, @ndellingwood, with the merge of:

* https://github.com/kokkos/kokkos/pull/6104

to Kokkos 'develop', can we merge this PR to Trilinos 'develop'?

**Comment by ndellingwood** [comment]
The kokkos-kernels changes are also merged https://github.com/kokkos/kokkos-kernels/pull/1817, thanks for working on this @bartlettroscoe 


> can we merge this PR to Trilinos 'develop'

@crtrott the kokkos and kokkos-kernels PRs corresponding to changes in this PR have merged to the resp. directories, it seems all blockers have been addressed, any final concerns before merge?

**Comment by bartlettroscoe** [comment]
@crtrott, @rppawlo, @ndellingwood, @ibaned, @sebrowne, @jwillenbring, @jhux2

Can we find someone to finish the review and approval of this PR?  This is currently holding up the rest of the work.  (And we need to know if customers will have any major problems with this change.)

**Comment by ndellingwood** [comment]
@bartlettroscoe I haven't heard back on any further concerns for the kokkos and kokkos-kernels changes, and the corresponding PRs have merged to their resp. repositories, so I'll go ahead and approve for the kokkos and and kokkos-kernels updates (I can't vouch for other packages)

**Review by ndellingwood — APPROVED** [review]
Kokkos and KokkosKernels changes were reviewed, approved, and have merged to their resp. repos

**Comment by bartlettroscoe** [comment]
> Kokkos and KokkosKernels changes were reviewed, approved, and have merged to their resp. repos

Thanks @ndellingwood!

> @bartlettroscoe I haven't heard back on any further concerns for the kokkos and kokkos-kernels changes, and the corresponding PRs have merged to their resp. repositories, so I'll go ahead and approve for the kokkos and and kokkos-kernels updates (I can't vouch for other packages)

I will go ahead and merge and then we can address any lingering issues in a follow-on PR.  Mostly we need to see if close SNL customers will have anything break from these changes.

**Comment by bartlettroscoe** [comment]
NOTE: I just noticed that:

* [gitlab-ex.sandia.gov/trilinos-devops-consolidation/code/GenConfig/-/merge_requests/92](https://gitlab-ex.sandia.gov/trilinos-devops-consolidation/code/GenConfig/-/merge_requests/92) (test name changes)
* [gitlab-ex.sandia.gov/trilinos-project/son-ini-files/-/merge_requests/3](https://gitlab-ex.sandia.gov/trilinos-project/son-ini-files/-/merge_requests/3) (test name changes)

are just changing the names of tests of the GenConfig tool and don't need to match what is currently in Trilinos.

Therefore, it is not needed to merge these MRs to the GenConfig system.

**Pull request merged**
